src/persist-client/src/internal/machine.rs

```diff
@@ -1070,6 +1085,8 @@ where
                                 .state_versions
                                 .fetch_current_state(&self.state.shard_id, all_live_diffs)
                                 .await
+                                .expect("shard codecs should not change")
+                                .check::<K, V, D>()
```

Comment:
Maybe the only downside of this approach is while the wrapper enforces that there _was_ a check, nobody enforces that the types that were checked were the types of the corresponding `Machine`. (eg. a call to `.check::<K, (), D>()` here would typecheck and mostly work in practise but still be wrong).

Have you considered, and/or how would you feel about, flipping the wrapper around so that the regular `State` struct is the untyped thing and we have some additional typed wrapper:

```rust
struct TypedState<K, V, T, D> {
  state: State<T>,
  pub(crate) _phantom: PhantomData<fn() -> (K, V, D)>,
}

impl<K, V, T, D> TypedState<K, V, T, D> {
  fn check(state: State<T>) -> Result<Self, _> { ... }
}
```

So the types carry the proof not only that the type has been checked, but match the specific types that match the caller.

Reply:
yeah, that's a great point and a _lovely_ solution! I think I'd counter-propose that we optimize the naming for the common case and do State vs UntypedState instead (which resolves my naming question in "Tips for reviewer")

this does mean all the `self.state` calls in Machine become something like `self.state.state`, but we can probably fix it with Deref+DerefMut and it probably goes away entirely once we have my Applier PR

Reply:
I don't feel that strongly about the naming; optimizing for the common case sounds fine to me!